subsys/storage/flash_buffered_write/fbw.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2017, 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME FBW
+#define LOG_LEVEL CONFIG_FBW_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_FBW_LOG_LEVEL);
+
+#include <zephyr/types.h>
+#include <string.h>
+#include <drivers/flash.h>
+
+#include <storage/fbw.h>
+
+#ifdef CONFIG_FBW_ERASE
+
+int fbw_erase(struct fbw_ctx *ctx, off_t off)
+{
+	int rc;
+	struct flash_pages_info page;
+
+	rc = flash_get_page_info_by_offs(ctx->fdev, off, &page);
+	if (rc != 0) {
+		LOG_ERR("Error %d while getting page info", rc);
+		return rc;
+	}
+
+	if (ctx->last_erased_page_start_offset == page.start_offset) {
+		return 0;
+	}
+
+	ctx->last_erased_page_start_offset = page.start_offset;
+	LOG_INF("Erasing page at offset 0x%08lx", (long)page.start_offset);
+
+	flash_write_protection_set(ctx->fdev, false);
+	rc = flash_erase(ctx->fdev, page.start_offset, page.size);
+	flash_write_protection_set(ctx->fdev, true);
+
+	if (rc != 0) {
+		LOG_ERR("Error %d while erasing page", rc);
+	}
+
+	return rc;
+}
+
+#endif /* CONFIG_FBW_ERASE */
+
+static int flash_sync(struct fbw_ctx *ctx)
+{
+	int rc = 0;
+	size_t write_addr = ctx->offset + ctx->bytes_written;
+
+
+	if (IS_ENABLED(CONFIG_FBW_ERASE)) {
+		rc = fbw_erase(ctx, write_addr + ctx->buf_bytes);
```

Comment:
I think this should be `write_addr + ctx->buf_bytes - 1U`, with a check somewhere that `ctx->buf_bytes` must be positive.  Otherwise you're erasing the page containing the byte that would be written after the ctx buffer is written, which may not exist and so may return an error, preventing the ctx buffer from being written even if it would be entirely stored in the last page.

Reply:
You are correct. I added a unit test triggering this bug in addition to your fix. Since buf_bytes is type `size_t` I simply return from the function if it is `0`. The caller of this function also ensures that `buf_bytes > 0` before calling it.